### ISSUE TITLE
New dart.mojom structure makes mocking difficult

### ISSUE THREAD

**Issue by abarth** [domokit/mojo#786]
I'd like to interpose a mock implementation of a service when someone connects to a service.  Previously, I mutated the `impl` property of the mojom.FooProxy objects.  With the new scheme, it's very awkward to interpose a mock.

Happy to brainstorm about possible solutions.

/cc @zanderso

**Comment by abarth** [comment]
See https://github.com/flutter/flutter/pull/4258 for the approach I was able to hack together.

**Comment by zanderso** [comment]
I think I should be able to restore something similar to what you were doing by allowing `impl` to be set in the generated ProxyControl here:

https://github.com/domokit/mojo/blob/master/mojo/public/tools/bindings/generators/dart_templates/interface_definition.tmpl#L151

and then checking if there's an `impl` to defer to in the generated Proxy. So, instead of writing `proxy.ptr = mock`, you'd write `proxy.impl = mock`.

**Comment by abarth** [comment]
Thanks.  That should work.

**Comment by j9brown** [comment]
If mocking is going to be a common use-case, we might want to offer a
supported and easily discoverable method for a setting it up.

**Comment by krisgiesing** [comment]
+1

**Comment by zanderso** [comment]
https://codereview.chromium.org/2024193002/

I've also added a `fromMock` factory constructor to the generated Interface class. So you can say, e.g.:

```dart
var echo = new EchoServiceInterface.fromMock(new EchoServiceMock());
```

as in the test here:

https://codereview.chromium.org/2024193002/diff/1/mojo/dart/apptests/dart_apptests/lib/src/echo_apptests.dart

**Comment by abarth** [comment]
I believe what I actually need to mock is an EchoServiceProxy.

**Comment by zanderso** [comment]
I've been trying to hide Proxies and Stubs to prefer more Mojo-y terminology. You can also do:

```dart
EchoServiceProxy echo = new EchoServiceProxy.fromMock(new EchoServiceMock());
```

As with the example above, this makes a Proxy whose calls are serviced by `EchoServiceMock()` instead of a remote service. Unless I've misunderstood https://github.com/flutter/flutter/pull/4258, this is what you were doing.

**Comment by pylaligand** [comment]
Another thing that's hard to test right now is the `ProxyControl`. In particular, I would like to be able to set its `errorFuture`, but right now I can't as the `ctrl` object's type is hidden.

This is what my mocks typically looked like:
```dart
class MockMyServiceProxy extends TypedMock implements MyServiceProxy {

  final Completer _completer = new Completer.sync();  

  MockMyServiceProxy() {
    when(close()).thenReturn(new Future.value());
  }

  @override
  Future get errorFuture => _completer.future;

  disconnect() => _completer.complete();
}
```

This way I could test the disconnection logic very easily. But now I can't see how to set `errorFuture` since I can't even mock `ctrl`. Adding a setter for `errorFuture` won't work either since I'm implementing the proxy class and not extending it.

**Comment by zanderso** [comment]
@pylaligand this sounds like a separate different problem from @abarth's. Could you open a separate issue?

**Comment by zanderso** [comment]
This change is now available in version 0.2.28 of the mojo_sdk pub package.
